### ISSUE TITLE
Player does not recover from segment requests that return a 404

### ISSUE THREAD

**Issue by BCjwhisenant** [google/ExoPlayer#537]
I'm testing with the ExoPlayer demo app, and have a few test manifests that I've added to the HLS Samples array:
```
new Sample("5 segments, should be normal playback", "http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/test-5-segments.m3u8", DemoUtil.TYPE_HLS),
new Sample("5 segments, first missing, player should recover and continue playing", "http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/test-first-segment-404.m3u8", DemoUtil.TYPE_HLS),
new Sample("5 segments, last missing, player should recover and complete", "http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/test-last-segment-404.m3u8", DemoUtil.TYPE_HLS),
```

Each test manifest includes 5 10-second ts segments.

When I play the first Sample, it plays to completion as expected.
When I play the second sample, a 404 is returned for the first segment request, as expected, but the player does not recover and play any subsequent segments.
When I play the third sample, a 404 is returned for last segment request, as expected, but in this case playback stops at about 10 seconds.

So far, I've seen this behavior on devices running 4.4.2 (Asus K011 and Samsung GT-N7100)), and 5.0.1 (Motorola Nexus 6).

Here's the logcat from the test where the first segment request returns a 404:
```
06-10 16:24:17.397    4132-4132/com.google.android.exoplayer.demo V/ContextImpl﹕ ----- packageName = com.google.android.exoplayer.demo is NOT LOCKED -----
06-10 16:24:17.427    4132-4132/com.google.android.exoplayer.demo I/ExoPlayerImpl﹕ Init 1.3.3
06-10 16:24:17.427    4132-4132/com.google.android.exoplayer.demo D/EventLogger﹕ start [0]
06-10 16:24:17.427    4132-4132/com.google.android.exoplayer.demo D/EventLogger﹕ state [0.00, false, P]
06-10 16:24:17.427    4132-4132/com.google.android.exoplayer.demo D/EventLogger﹕ state [0.00, true, P]
06-10 16:24:17.457    4132-4132/com.google.android.exoplayer.demo D/﹕ Pixel Format : GGL_PIXEL_FORMAT_RGBA_8888
06-10 16:24:17.527    4132-4132/com.google.android.exoplayer.demo D/﹕ Surface destroy: ANDROID_NATIVE_WINDOW_MAGIC
06-10 16:24:18.017    4132-5900/com.google.android.exoplayer.demo D/dalvikvm﹕ GC_FOR_ALLOC freed 3332K, 31% free 7678K/11060K, paused 18ms, total 18ms
06-10 16:24:18.017    4132-5904/com.google.android.exoplayer.demo I/OMXClient﹕ Using client-side OMX mux.
06-10 16:24:18.027    4132-4132/com.google.android.exoplayer.demo D/EventLogger﹕ state [0.59, true, B]
06-10 16:24:18.027    4132-4132/com.google.android.exoplayer.demo D/EventLogger﹕ videoFormat [0.59, 0, 0]
06-10 16:24:18.027    4132-5904/com.google.android.exoplayer.demo E/ACodec﹕ [OMX.Intel.hw_vd.h264] storeMetaDataInBuffers failed w/ err -2147483648
06-10 16:24:18.037    4132-5912/com.google.android.exoplayer.demo I/OMXClient﹕ Using client-side OMX mux.
06-10 16:24:18.047    4132-4132/com.google.android.exoplayer.demo D/EventLogger﹕ decoderInitialized [0.61]
06-10 16:24:18.087    4132-4132/com.google.android.exoplayer.demo D/EventLogger﹕ videoSizeChanged [240, 180, 1.0]
06-10 16:24:18.197    4132-4132/com.google.android.exoplayer.demo D/EventLogger﹕ state [0.77, true, R]
06-10 16:24:35.067    4132-4132/com.google.android.exoplayer.demo D/﹕ Pixel Format : GGL_PIXEL_FORMAT_RGBA_8888
06-10 16:24:37.087    4132-4132/com.google.android.exoplayer.demo E/EventLogger﹕ internalError [19.66, loadError]
    com.google.android.exoplayer.upstream.HttpDataSource$InvalidResponseCodeException: Response code: 404
            at com.google.android.exoplayer.upstream.DefaultHttpDataSource.open(DefaultHttpDataSource.java:204)
            at com.google.android.exoplayer.upstream.DefaultUriDataSource.open(DefaultUriDataSource.java:136)
            at com.google.android.exoplayer.hls.TsChunk.load(TsChunk.java:102)
            at com.google.android.exoplayer.upstream.Loader$LoadTask.run(Loader.java:242)
            at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
            at java.lang.Thread.run(Thread.java:841)
06-10 16:24:37.197    4132-4132/com.google.android.exoplayer.demo E/EventLogger﹕ internalError [19.77, loadError]
    com.google.android.exoplayer.upstream.HttpDataSource$InvalidResponseCodeException: Response code: 404
            at com.google.android.exoplayer.upstream.DefaultHttpDataSource.open(DefaultHttpDataSource.java:204)
            at com.google.android.exoplayer.upstream.DefaultUriDataSource.open(DefaultUriDataSource.java:136)
            at com.google.android.exoplayer.hls.TsChunk.load(TsChunk.java:102)
            at com.google.android.exoplayer.upstream.Loader$LoadTask.run(Loader.java:242)
            at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
            at java.lang.Thread.run(Thread.java:841)
06-10 16:24:38.327    4132-4132/com.google.android.exoplayer.demo E/EventLogger﹕ internalError [20.90, loadError]
    com.google.android.exoplayer.upstream.HttpDataSource$InvalidResponseCodeException: Response code: 404
            at com.google.android.exoplayer.upstream.DefaultHttpDataSource.open(DefaultHttpDataSource.java:204)
            at com.google.android.exoplayer.upstream.DefaultUriDataSource.open(DefaultUriDataSource.java:136)
            at com.google.android.exoplayer.hls.TsChunk.load(TsChunk.java:102)
            at com.google.android.exoplayer.upstream.Loader$LoadTask.run(Loader.java:242)
            at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
            at java.lang.Thread.run(Thread.java:841)
06-10 16:24:40.477    4132-4132/com.google.android.exoplayer.demo E/EventLogger﹕ internalError [23.05, loadError]
    com.google.android.exoplayer.upstream.HttpDataSource$InvalidResponseCodeException: Response code: 404
            at com.google.android.exoplayer.upstream.DefaultHttpDataSource.open(DefaultHttpDataSource.java:204)
            at com.google.android.exoplayer.upstream.DefaultUriDataSource.open(DefaultUriDataSource.java:136)
            at com.google.android.exoplayer.hls.TsChunk.load(TsChunk.java:102)
            at com.google.android.exoplayer.upstream.Loader$LoadTask.run(Loader.java:242)
            at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
            at java.lang.Thread.run(Thread.java:841)
06-10 16:24:40.487    4132-5895/com.google.android.exoplayer.demo E/ExoPlayerImplInternal﹕ Internal track renderer error.
    com.google.android.exoplayer.ExoPlaybackException: com.google.android.exoplayer.upstream.HttpDataSource$InvalidResponseCodeException: Response code: 404
            at com.google.android.exoplayer.metadata.MetadataTrackRenderer.doSomeWork(MetadataTrackRenderer.java:150)
            at com.google.android.exoplayer.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:401)
            at com.google.android.exoplayer.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:203)
            at android.os.Handler.dispatchMessage(Handler.java:98)
            at android.os.Looper.loop(Looper.java:149)
            at android.os.HandlerThread.run(HandlerThread.java:61)
            at com.google.android.exoplayer.util.PriorityHandlerThread.run(PriorityHandlerThread.java:40)
     Caused by: com.google.android.exoplayer.upstream.HttpDataSource$InvalidResponseCodeException: Response code: 404
            at com.google.android.exoplayer.upstream.DefaultHttpDataSource.open(DefaultHttpDataSource.java:204)
            at com.google.android.exoplayer.upstream.DefaultUriDataSource.open(DefaultUriDataSource.java:136)
            at com.google.android.exoplayer.hls.TsChunk.load(TsChunk.java:102)
            at com.google.android.exoplayer.upstream.Loader$LoadTask.run(Loader.java:242)
            at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
            at java.lang.Thread.run(Thread.java:841)
06-10 16:24:40.487    4132-4132/com.google.android.exoplayer.demo E/EventLogger﹕ playerFailed [23.05]
    com.google.android.exoplayer.ExoPlaybackException: com.google.android.exoplayer.upstream.HttpDataSource$InvalidResponseCodeException: Response code: 404
            at com.google.android.exoplayer.metadata.MetadataTrackRenderer.doSomeWork(MetadataTrackRenderer.java:150)
            at com.google.android.exoplayer.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:401)
            at com.google.android.exoplayer.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:203)
            at android.os.Handler.dispatchMessage(Handler.java:98)
            at android.os.Looper.loop(Looper.java:149)
            at android.os.HandlerThread.run(HandlerThread.java:61)
            at com.google.android.exoplayer.util.PriorityHandlerThread.run(PriorityHandlerThread.java:40)
     Caused by: com.google.android.exoplayer.upstream.HttpDataSource$InvalidResponseCodeException: Response code: 404
            at com.google.android.exoplayer.upstream.DefaultHttpDataSource.open(DefaultHttpDataSource.java:204)
            at com.google.android.exoplayer.upstream.DefaultUriDataSource.open(DefaultUriDataSource.java:136)
            at com.google.android.exoplayer.hls.TsChunk.load(TsChunk.java:102)
            at com.google.android.exoplayer.upstream.Loader$LoadTask.run(Loader.java:242)
            at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
            at java.lang.Thread.run(Thread.java:841)
06-10 16:24:40.517    4132-4132/com.google.android.exoplayer.demo D/EventLogger﹕ state [23.09, true, I]
```

**Comment by ojw28** [comment]
We survive provided the segment is available in one of the possible variants. In this case there is only one possible variant, and so we fail.

I'm not sure we're particularly interested in making playback robust against these kind of cases. Streaming services should just not list chunks in the playlist if they're unable to serve them. If you're going to be delivering a good quality service this kind of case shouldn't occur.

**Comment by BCjwhisenant** [comment]
Hello,

I have a couple of tests with two variants each:
http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-gear1-seg6-404.m3u8
http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-gear2-seg6-404.m3u8

In each of these tests, one of the variants has all of its segments (there are 11 segments total in each test). I've included two tests, since I couldn't guarantee which variant would be playing (in my testing, it was always the "gear2" variant, but if your bandwidth is different, you may load the "gear1" variant instead).

In my testing tonight, I'm seeing that rather than switch to the other variant in the playlist, the player stops playback at about 0:25, when it receives a 404 for segment 6 in the variant that's currently loaded.

I hope this is a more useful example of the problem.

**Comment by ojw28** [comment]
Thanks! It turns out we're only blacklisting if the playlist request fails, not if an individual chunk request fails. I'll fix this.

**Comment by ojw28** [comment]
This should work now. Please check though!

**Comment by BCjwhisenant** [comment]
I've verified on the dev branch that the player recovers if the segment request fails. Thanks so much for this.

You mentioned above that the blacklisting was previously only applied to playlists. It seems to work only if the unavailable playlist is not the first-listed playlist in the manifest.

I have a couple of tests here:
http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-gear1-playlist-404.m3u8
and
http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-gear2-playlist-404.m3u8

In each case, the gear1 playlist is listed first.

In the "bipbop-test-gear1-playlist-404" case, the player never recovers to attempt to request the gear2 playlist. In the "bipbop-test-gear2-playlist-404" case, though, since the gear1 playlist is available, the player is able to blacklist the gear2 playlist as expected.

I can open this into a separate bug if preferred.

**Comment by ojw28** [comment]
The indexing in that class is a bit messed up as well.

**Comment by ojw28** [comment]
I'm surprised this ever worked (it probably works by chance, mainly). I've sanitized the whole class in 51a8635ba2fa470ef93ccf6d48a3de0fadb2150f. I believe your test cases should pass now. Please verify.

**Comment by BCjwhisenant** [comment]
In one case that I have, it still fails:
http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-gear1-seg1-404.m3u8
In this case, the first segment of the first-listed variant is unavailable. Even though the player downloads segments (including the first segment) for the available playlist, it never recovers to play them.

**Comment by BCjwhisenant** [comment]
Otherwise, my testing has been successful. Thanks!

**Comment by ojw28** [comment]
The final case was a problem further downstream, and should be fixed by 203f3ab7321a21900cc842ba9f0ef17fe29a1607. Thanks!

**Comment by BCjwhisenant** [comment]
Verified that all of the following cases now work:

Unavailable gear1 playlist (player recovers and initiates playback with gear 2)
Unavailable gear2 playlist (player recovers and continues playback with gear 1)
Unavailable gear1 segment 1 (player recovers and initiates playback with gear 2)
Unavailable gear1 segment 6 (player has switched to gear2 by segment 6, and so it continues playback with gear 2 uninterrupted)
Unavailable gear2 segment 1 (since the player has already initiated playback with gear1, it's able to seamlessly switch to gear 2 and continue)
Unavailable gear2 segment 6 (player recovers and resumes playback with gear 1 segment 6)

Thanks!

In testing, noticed that when the player loads, it loads the first two segments of gear1 (since gear1 is the first variant in the master playlist). As the first of these two segments is playing, the second segment from gear2 is requested, so that the player can switch to the best variant available (as expected).

Based on the observation that two segments are loaded from the first variant, I think I've uncovered a corner case that may be interesting. I'm happy to open this as a separate issue if desired.

In testing the missing-first-segment scenarios above, I also added two cases:
http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-gear1-seg2-404.m3u8
http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-gear2-seg2-404.m3u8

In the first case, playback continues uninterrupted, as expected, since the segment is available in the second variant.
However, in the second case, it appears that while the first segment of gear1 is playing, as soon as the second segment of gear2 is requested (which will ultimately return a 404), the second segment of gear1 is discarded (note that it hasn’t begun to play yet). After the request for gear2 segment 2 returns the 404, the player re-requests gear1 segment1, but unfortunately it is not able to play it.

It's a bit of an oddball-sounding case (even to me! :)), but I wanted to pass it along nonetheless.
Thanks again!

**Comment by ojw28** [comment]
Heh. I think there's a simple fix for that. I'll merge one on Monday, and have re-opened this issue to track that. There are also some issues around the value of previousTsChunk provided to HlsChunkSource.getChunkOperation in the case where a chunk needs to be replaced. It's currently providing the wrong chunk.

**Comment by ojw28** [comment]
Merged b8df8ecb47d838f4e3c184cbcbb64a397b1384a2 and 97aaee6d1505e52f0c502c19758ba59de27dc959. I didn't test them in detail.

Thanks for helping to iron these cases out, it's been very useful. If possible, could you:

- Verify again that all of the cases are now working as you expect them to?
- Provide descriptions of each test. Basically just like your list above, but this time including the additional cases that you added, and with the corresponding link in each case. We're working on playback tests that we can run automatically, and these cases seem like great candidates. If you could provide this list, we'll make sure to come up with equivalent tests for our test set (nb - we're planning on hosting all of the test content ourselves - you're not committing to keeping the links live or anything like that :)).

**Comment by BCjwhisenant** [comment]
My pleasure to help.
I'll put a list together this weekend.

On Fri, Jun 19, 2015 at 3:33 PM, ojw28 <notifications@github.com> wrote:

> Merged b8df8ec
> <https://github.com/google/ExoPlayer/commit/b8df8ecb47d838f4e3c184cbcbb64a397b1384a2>
> and 97aaee6
> <https://github.com/google/ExoPlayer/commit/97aaee6d1505e52f0c502c19758ba59de27dc959>.
> I didn't test them in detail.
>
> Thanks for helping to iron these cases out, it's been very useful. If
> possible, could you:
>
>    - Verify again that all of the cases are now working as you expect
>    them to?
>    - Provide descriptions of each test. Basically just like your list
>    above, but this time including the additional cases that you added, and
>    with the corresponding link in each case. We're working on playback tests
>    that we can run automatically, and these cases seem like great candidates.
>    If you could provide this list, we'll make sure to come up with equivalent
>    tests for our test set (nb - we're planning on hosting all of the test
>    content ourselves - you're not committing to keeping the links live or
>    anything like that :)).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/ExoPlayer/issues/537#issuecomment-113615324>.
>



-- 

Jim Whisenant

Senior Software Engineer in Test

Brightcove Inc. <http://www.brightcove.com/>

290 Congress St., 3rd Floor, Boston, MA 02110

P: 617 500 4947  F: 617 261 4830

**Comment by BCjwhisenant** [comment]
Here is a summary of the tests I've done, along with expected and actual results.

These test cases are based on the Apple BipBop HLS sample video. In order to have somewhat concise tests, these cases use a small subset of the actual video data, as detailed below.

Each test case consists of the following:
- A master manifest
- The master manifest lists two playlist manifests
- Each playlist manifest lists 11 segments

| Test Description | URL | Expected Result | Pass/Fail | Notes |
|----------------------|----------------------------------------------------|-------------------------------------------------|------------------|------------------|
All playlists and segments available | http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-all-segs-200.m3u8 | No errors or warnings should be thrown during loading or playback. The gear1 playlist should be initially selected, but the player should switch to gear2 if bandwidth allows. | PASS |
| Unavailable gear1 playlist | http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-gear1-playlist-404.m3u8 | Player should recover and initiate playback with gear 2 playlist; gear1 segments should not be requested for the remainder of playback, including seeking and replay | PASS |
| Unavailable gear2 playlist | http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-gear2-playlist-404.m3u8 | Player should continue playback with gear 1; gear2 segments should not be requested for the remainder of playback, including seeking and replay | PASS | on replay, the gear2 playlist is requested again, but ignored again |
| Unavailable gear1 fileSequence0 | http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-gear1-seg1-404.m3u8 | Player should recover and initiate playback with gear2; gear1 segments should not be requested for the remainder of playback, including seeking and replay | FAIL | Playback never started; instead, the demo app crashed |
| Unavailable gear1 fileSequence1 | http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-gear1-seg2-404.m3u8 | Since the Player loads two segments before starting playback, Player should recover and continue playback with gear2 segment 2; gear1 segments should not be requested for the remainder of playback | PASS |
| Unavailable gear1 fileSequence6 | http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-gear1-seg6-404.m3u8 | If Player has switched to gear2 by segment 6, it should continue playback with gear 2 uninterrupted. If it has not switched to gear2 yet, it should attempt to do so, as long as bandwidth permits | PASS |
| Unavailable gear2 fileSequence0 | http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-gear2-seg1-404.m3u8 | Since the player has already initiated playback with gear1 segment 1, and not attempted to load gear2’s missing segment, the gear2 playlist should not be blacklisted yet | FAIL ON REPLAY | After the player has switched to gear2, if you try to seek back to the beginning, the missing gear2 segment will be requested, and the app will crash |
| Unavailable gear2 fileSequence1 | http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-gear2-seg2-404.m3u8 | Player should initiate and continue playback with the gear1 playlist, including seeking and replay | PASS |
| Unavailable gear2 fileSequence6 | http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/bipbop-test-gear2-seg6-404.m3u8 | Player recovers and resumes playback with gear 1 segment 6, including seeking and replay | PASS | on replay, segments from gear2 are requested again, but the Player switches to gear1 at fileSequence6, as in the original playback |

Note that in the negative-path segment cases, the playlist manifests will have the same segment lists, that is, both playlists will list fileSequence0.ts, fileSequence1.ts, fileSequence2.ts, etc. I wanted to have as close to an apples-to-apples comparison in the playlist manifest contents as possible. To produce the 404 return codes, I actually removed the segments whose requests were expected to product those errors. Feel free to inspect the page at http://solutions.brightcove.com/jwhisenant/testpages/exoplayer/hls/bipbop/ for more information, or you can ping me as needed.

**Comment by ojw28** [comment]
All cases should now pass. Thanks.

**Comment by BCjwhisenant** [comment]
Indeed, all of the multi-variant cases pass. Thanks very much!

For the earlier single-variant cases, it does seem plausible that if an uploaded video is not of a high enough bitrate, then an audio-only variant could be transcoded, which would automatically create at least two variants.

Again, thanks a ton.
